pygac/l1c_processor.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2014, 2019 Pygac Developers
+
+# Author(s):
+
+#   Carlos Horn <carlos.horn@external.eumetsat.int>
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""l1c files processor
+
+"""
+import datetime
+import logging
+
+from pygac.reader import ReaderError
+from pygac.gac_klm import GACKLMReader
+from pygac.gac_pod import GACPODReader
+from pygac.lac_klm import LACKLMReader
+from pygac.lac_pod import LACPODReader
+from pygac.utils import file_opener
+
+
+LOG = logging.getLogger(__name__)
+
+_Readers = [GACKLMReader, LACKLMReader, GACPODReader, LACPODReader]
+
+
+def get_reader(filename, fileobj=None):
+    """Read the GAC/LAC KLM/POD data.
+
+        Args:
+            filename (str): Path to GAC/LAC file
+            fileobj: An open file object to read from. (optional)
+        """
+    found_reader = False
+    with file_opener(fileobj or filename) as open_file:
+        for i, Reader in enumerate(_Readers):
+            try:
+                reader = Reader.fromfile(filename, fileobj=open_file)
+                found_reader = True
+                index = i
+            except ReaderError as exception:
+                LOG.debug("%s failed to read the file! %s"
+                          % (Reader.__name__, repr(exception)))
+            finally:
+                open_file.seek(0)
+            if found_reader:
+                break
+    if not found_reader:
+        raise ValueError('Unable to read the file "%s"' % filename)
+    # Move the Reader in front of _Readers. Chance is high that the
+    # next file is of the same kind.
+    _Readers.insert(0, _Readers.pop(index))
+    return reader
```

Comment:
I'm happy this function got added, it will make it easier to use pygac from other libraries. However, if I read this correctly, the entire file is read and then the file pointer is put back to the beginning of the file.

I would like to avoid having to read the full file for this, because if we want to sort out files in a directory full of gac/lac files, this will might slow down the search.

Maybe `_validate_header` should be called here instead of `fromfile` ?

Reply:
Welcome back from vacation!
It appears like it will read the entire file, but it actually does only read the header. The header validation is performed in `reader.read` just after reading the header. Rewinding the file and re-reading the header with the next reader should not (much) affect the file system, because python uses an internal buffer for open files which should still contain the relevant bytes.

Reply:
Thanks !
I'm not following: `fromfile` calls `read`, in which I can see:
```python
                self.scans = np.frombuffer(
                    fd_.read(count*self.scanline_type.itemsize),
                    dtype=self.scanline_type,
                    count=count)
```
or am I missing something ?

Reply:
So if the header matches, the file is read entirely, right ?

Reply:
Yes right, if a file is called with the wrong reader, the `read` method would raise the `ReaderError` just after reading the header and return to the loop in `get_reader`. Only if the header passes the validation, it reads the scans. The important thing is that `fd_.read` is never called without an argument, so it never reads more than necessary before validation.

Reply:
Ok, so something needs to be fixed here: either the function name, or the what is happening inside. `get_reader` for me means I provide a file and get back a reader class I can instantiate (which would be really useful in eg satpy)...

Reply:
I see. You want a `get_Reader` function to get the class. I only head the instance in mind...
Should be easy to implement. I would split the `read` into two parts, which are `read_header` and `read_scans`. Maybe the `read_header` should be a class method which returns the ars/tbm_header and the header, or raises the `ReaderError` if the header does not fit.
Then we can use this class method in `get_Reader` to select the appropriate reader and return the class.

Reply:
That would be great !

Reply:
Now, the reader has the classmethod `XxxReader.can_read` which returns `True` if the reader understands the header otherwise `False`. This can be used for checking files against a given reader as done in `get_reader_class`. 
I kept the `reader.fromfile` method as it is, because to me it appears natural to have an alternative constructor which initializes the whole instance from the given input in analogy to  `datetime.datetime.fromisoformat` for example.